xdsl/transforms/convert_ptr_to_llvm.py

```
@@ -0,0 +1,274 @@
+            llvm.GEPOp.from_mixed_indices(
+                cast_addr_op.results[0],
+                [cast_offset_op.result],
+                pointee_type=builtin.f32,
+            )
+        )
+
+
+class ReconcileUnrealizedPtrCasts(RewritePattern):
+    """
+    Eliminates two types of unrealized casts:
+    - `llvm.ptr` -> `llvm.ptr`
+    - `llvm.ptr` -> `ptr_xdsl.ptr` -> `llvm.ptr`
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: builtin.UnrealizedConversionCastOp, rewriter: PatternRewriter, /
+    ):
+        # preconditions
+        if (
+            len(op.inputs) != 1
+            or len(op.outputs) != 1
+            or not isinstance(op.inputs[0].type, llvm.LLVMPointerType)
+        ):
+            return
+
+        # erase llvm.ptr -> llvm.ptr
+        if isinstance(op.outputs[0].type, llvm.LLVMPointerType):
+            op.outputs[0].replace_by(op.inputs[0])
+            rewriter.erase_matched_op()
+            return
+
+        if not isinstance(op.outputs[0].type, ptr.PtrType):
+            return
+
+        # erase llvm.ptr -> ptr_xdsl.ptr -> llvm.ptr cast pairs
+        uses = tuple(use for use in op.outputs[0].uses)
+        for use in uses:
+            if (
+                isinstance(use.operation, builtin.UnrealizedConversionCastOp)
+                and isinstance(use.operation.inputs[0].type, ptr.PtrType)
+                and isinstance(use.operation.outputs[0].type, llvm.LLVMPointerType)
+            ):
+                use.operation.outputs[0].replace_by(op.inputs[0])
+                rewriter.erase_op(use.operation)
+
+        # erase this cast entirely if no uses are remaining
+        if op.outputs[0].uses:
+            return
+
+        rewriter.erase_op(op)
+
+
+@dataclass
+class ConvertTypeOffsetOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: ptr.TypeOffsetOp, rewriter: PatternRewriter, /):
+        # TODO: replace with memoized dict lookup
```

we have a helper for this, it should just be `op.elem_type.size` if it's a `FixedBitwidthType`, we might want to raise an error otherwise.

---

Btw I [merged](https://github.com/xdslproject/xdsl/commit/23d2a0eda1d7f2d8c15619898b23c2a2350b6f5f) the type offset conversion pass so maybe you can just use it in this case